staging/src/k8s.io/apiserver/pkg/util/x509metrics/missing_san.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package x509metrics
+
+import (
+	"crypto/x509"
+	"errors"
+	"net/http"
+	"strings"
+
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/component-base/metrics"
+)
+
+var _ utilnet.RoundTripperWrapper = &x509MissingSANErrorMetricsRTWrapper{}
+
+type x509MissingSANErrorMetricsRTWrapper struct {
+	rt http.RoundTripper
+
+	counter *metrics.CounterVec
+}
+
+// NewMissingSANRoundTripperWrapperConstructor returns a RoundTripper wrapper that's usable
+// within ClientConfig.Wrap that increases the `metricCounter` whenever:
+// 1. we get a x509.HostnameError with string `x509: certificate relies on legacy Common Name field`
+//    which indicates an error caused by the deprecation of Common Name field when veryfing remote
+//    hostname
+// 2. the server certificate in response contains no SAN. This indicates that this binary run
+//    with the GODEBUG=x509ignoreCN=0 in env
+func NewMissingSANRoundTripperWrapperConstructor(metricCounter *metrics.CounterVec) func(rt http.RoundTripper) http.RoundTripper {
+	return func(rt http.RoundTripper) http.RoundTripper {
+		return &x509MissingSANErrorMetricsRTWrapper{
+			rt:      rt,
+			counter: metricCounter,
+		}
+	}
+}
+
+func (w *x509MissingSANErrorMetricsRTWrapper) RoundTrip(req *http.Request) (*http.Response, error) {
+	resp, err := w.rt.RoundTrip(req)
+
+	if err != nil && errors.As(err, &x509.HostnameError{}) {
+		// Run w/o GODEBUG=x509ignoreCN=0
+		if errMsg := err.Error(); strings.Contains(errMsg, "x509: certificate relies on legacy Common Name field") {
```

Comment:
is there a test that would catch if the golang error string changed in a way this would no longer match?

Reply:
I think unit tests are by default running without the environment variable so the unit-test additions in
`staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy_test.go` and `staging/src/k8s.io/apiserver/pkg/util/webhook/webhook_test.go` should serve for that purpose, but that would make the other branch of this function untetested.

I'll add unit tests to this package as I should have done anyway.

Reply:
... but that can't be done because we can't dynamically change the behavior of the crypto library, I completely forgot